File: LevelDBTests/LevelDBTests.m
//
//  LevelDBTests.m
//  LevelDB
//
//  Copyright (c) 2015 Pyry Jahkola. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <LevelDB/LevelDB.h>

@implementation NSString (LevelDBTests)
- (NSData *)ldb_UTF8Data
{
    return [self dataUsingEncoding:NSUTF8StringEncoding];
}
@end

@implementation NSData (LevelDBTests)
- (NSString *)ldb_UTF8String
{
    return [[NSString alloc] initWithData:self encoding:NSUTF8StringEncoding];
}
@end

@interface LevelDBTests : XCTestCase
@property NSString *path;
@end

@implementation LevelDBTests

- (void)setUp {
    [super setUp];
    NSString *unique = [NSProcessInfo processInfo].globallyUniqueString;
    self.path = [NSTemporaryDirectory() stringByAppendingPathComponent:unique];
}

- (void)tearDown {
    [LDBDatabase destroyDatabaseAtPath:self.path error:nil];
    NSAssert(![[NSFileManager defaultManager] fileExistsAtPath:self.path], @"");
    [super tearDown];
}

- (void)testExample {
    // This is an example of a functional test case.
    LDBDatabase *db = [LDBDatabase new];
    
    if (YES) {
        LDBSnapshot *empty = [db snapshot];
        for (LDBEnumerator *e = empty.enumerator; e.isValid; [e step]) {
            XCTFail(@"expected empty snapshot, got (%@, %@)", e.key, e.value);
        }
    }
    
    if (YES) {
        db[@"b".ldb_UTF8Data] = @"B".ldb_UTF8Data;
        db[@"d".ldb_UTF8Data] = @"D".ldb_UTF8Data;
        db[@"e".ldb_UTF8Data] = @"E".ldb_UTF8Data;
        db[@"d".ldb_UTF8Data] = nil;

        LDBSnapshot *snap = [db snapshot];
        XCTAssertEqualObjects(snap.enumerator.allObjects, (@[
            @[@"b".ldb_UTF8Data, @"B".ldb_UTF8Data],
            @[@"e".ldb_UTF8Data, @"E".ldb_UTF8Data],
        ]));
    }
        
    if (YES) {
        LDBWriteBatch *batch = [LDBWriteBatch new];
        batch[@"b".ldb_UTF8Data] = @"!".ldb_UTF8Data;
        batch[@"a".ldb_UTF8Data] = @"A".ldb_UTF8Data;
        batch[@"e".ldb_UTF8Data] = nil;
        batch[@"c".ldb_UTF8Data] = @"C".ldb_UTF8Data;
        NSError *error;
        BOOL ok = [db write:batch sync:NO error:&error];
        XCTAssert(ok);
        XCTAssertNil(error);
    }
    
    if (YES) {
        LDBSnapshot *snap = [db snapshot];
        XCTAssertEqualObjects(snap.enumerator.allObjects, (@[
            @[@"a".ldb_UTF8Data, @"A".ldb_UTF8Data],
            @[@"b".ldb_UTF8Data, @"!".ldb_UTF8Data],
            @[@"c".ldb_UTF8Data, @"C".ldb_UTF8Data],
        ]));
    }
}

- (void)testDemo
{
    LDBDatabase *database = [[LDBDatabase alloc] initWithPath:self.path];

    NSData *key1 = @"foo".ldb_UTF8Data;
    NSData *val1 = @"FOO".ldb_UTF8Data;
    database[key1] = val1;

    NSCAssert([database[key1] isEqual:val1], @"");

    LDBWriteBatch *batch = [LDBWriteBatch new];
    for (NSString *k in @[@"aha", @"bar", @"baz", @"boo", @"xyz"]) {
        batch[k.ldb_UTF8Data] =
            [k uppercaseString].ldb_UTF8Data;
    }
    NSError *error;
    if (![database write:batch sync:NO error:&error]) {
        NSLog(@"batch write failed: %@", error);
    } else {
        LDBSnapshot *snapshot = database.snapshot;

        NSLog(@"snapshot contents:");
        for (LDBEnumerator *e = snapshot.enumerator; e.isValid; [e step]) {
            NSLog(@"  - %@: %@", e.key.ldb_UTF8String, e.value.ldb_UTF8String);
        }
        XCTAssertEqualObjects(snapshot.enumerator.allObjects, (@[
            @[@"aha".ldb_UTF8Data, @"AHA".ldb_UTF8Data],
            @[@"bar".ldb_UTF8Data, @"BAR".ldb_UTF8Data],
            @[@"baz".ldb_UTF8Data, @"BAZ".ldb_UTF8Data],
            @[@"boo".ldb_UTF8Data, @"BOO".ldb_UTF8Data],
            @[@"foo".ldb_UTF8Data, @"FOO".ldb_UTF8Data],
            @[@"xyz".ldb_UTF8Data, @"XYZ".ldb_UTF8Data],
        ]));

        LDBSnapshot *clamped = [snapshot prefixed:@"ba".ldb_UTF8Data];
        NSLog(@"clamped snapshot contents:");
        for (NSArray *pair in clamped.enumerator) {
            NSLog(@"  - %@: %@", [pair[0] ldb_UTF8String], [pair[1] ldb_UTF8String]);
        }
        XCTAssertEqualObjects(clamped.enumerator.allObjects, (@[
            @[@"r".ldb_UTF8Data, @"BAR".ldb_UTF8Data],
            @[@"z".ldb_UTF8Data, @"BAZ".ldb_UTF8Data],
        ]));
    }
}

@end

File: LevelDBTests/LevelDBTests.swift
//
//  LevelDBTests.swift
//  LevelDBTests
//
//  Created by Pyry Jahkola on 26.01.2015.
//  Copyright (c) 2015 Pyrtsa. All rights reserved.
//

import Foundation
import XCTest
import LevelDB

class LevelDBTests: XCTestCase {

    var path = ""
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
        path = NSTemporaryDirectory().stringByAppendingPathComponent(NSProcessInfo.processInfo().globallyUniqueString)
        // NSLog("using temp path: %@", path)
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        LDBDatabase.destroyDatabaseAtPath(path, error: nil)
        assert(!NSFileManager.defaultManager().fileExistsAtPath(path))
        super.tearDown()
    }
    
    func testInMemory() {
        let db: LDBDatabase = LDBDatabase()
        XCTAssertNil(db[NSData()])
        db[NSData()] = NSData()
        XCTAssertNotNil(db[NSData()])
        db[NSData()] = nil
        XCTAssertNil(db[NSData()])
    }
    
    func testOnDisk() {
        let maybeDb: LDBDatabase? = LDBDatabase(path)
        XCTAssertNotNil(maybeDb)
        
        if maybeDb == nil { return }
        let db = maybeDb!
        XCTAssertNil(db[NSData()])
        
        db[NSData()] = NSData()
        XCTAssertNotNil(db[NSData()])
        
        db[NSData()] = nil
        XCTAssertNil(db[NSData()])
    }
    
//    func testStringDatabase() {
//    
//        let db = Database<String, String>()
//        
//        db["foo"] = "bar"
//        
//        let byteValued = db.cast() as Database<String, NSData>
//        let byteKeyed = db.cast() as Database<NSData, String>
//        let byteDb = db.cast() as Database<NSData, NSData>
//
//        XCTAssertEqual(db["foo"],             Optional("bar"))
//        XCTAssertEqual(byteValued["foo"],     Optional("bar".UTF8))
//        XCTAssertEqual(byteKeyed["foo".UTF8], Optional("bar"))
//        XCTAssertEqual(byteDb["foo".UTF8],    Optional("bar".UTF8))
//
//        db["foo"] = nil
//    
//        XCTAssertNil(db["foo"])
//
//    }
//    
//    func testWriteBatch() {
//    
//        let batch = WriteBatch<String, String>()
//        
//        batch.put("foo", "bar")
//        batch.delete("foo")
//        
//        XCTAssertEqual(batch.diff, [("foo", nil)])
//
//        batch.put("qux", "abc")
//        batch.delete("def")
//        batch.delete("bar")
//        batch.put("foo", "def")
//
//        XCTAssertEqual(batch.diff, [("bar", nil),
//                                    ("def", nil),
//                                    ("foo", "def"),
//                                    ("qux", "abc")])
//    
//        let db1 = Database<String, String>()
//        
//        db1["bar"] = "ghi"
//        db1["baz"] = "jkl"
//        
//        XCTAssertEqual(db1["bar"], "ghi")
//        XCTAssertEqual(db1["baz"], "jkl")
//        
//        XCTAssertNil(db1.write(batch).error)
//        
//        XCTAssertEqual(db1["bar"], nil)
//        XCTAssertEqual(db1["baz"], "jkl")
//        XCTAssertEqual(db1["def"], nil)
//        XCTAssertEqual(db1["foo"], "def")
//        XCTAssertEqual(db1["qux"], "abc")
//
//        let db2 = Database<String, String>(path)!
//        
//        db2["def"] = "ghi"
//        db2["baz"] = "jkl"
//        
//        XCTAssertEqual(db2["def"], "ghi")
//        XCTAssertEqual(db2["baz"], "jkl")
//        
//        XCTAssertNil(db2.write(batch).error)
//        
//        XCTAssertEqual(db2["bar"], nil)
//        XCTAssertEqual(db2["baz"], "jkl")
//        XCTAssertEqual(db2["def"], nil)
//        XCTAssertEqual(db2["foo"], "def")
//        XCTAssertEqual(db2["qux"], "abc")
//        
//    }
    
    func testOpenFailures() {
//        XCTAssertNotNil(Database<String, String>.open(path).error, "should fail with `createIfMissing: false`")
//        XCTAssertNotNil(Database<String, String>.open(path, createIfMissing: true).value, "should succeed with `createIfMissing: true`")
//        XCTAssertNotNil(Database<String, String>.open(path, errorIfExists: true).error, "should fail with `errorIfExists: true`")

        if true {
            var error: NSError?
            XCTAssertNil(LDBDatabase(path: path, error: &error))
            XCTAssertNotNil(error, "should fail with `createIfMissing: false`")
        }
        if true {
            var error: NSError?
            XCTAssertNotNil(LDBDatabase(path: path, error: &error, createIfMissing: true))
            XCTAssertNil(error, "should succeed with `createIfMissing: true`")
        }
        if true {
            var error: NSError?
            XCTAssertNil(LDBDatabase(path: path, error: &error, errorIfExists: true))
            XCTAssertNotNil(error, "should fail with `errorIfExists: true`")
        }
    }
    
    func testFilterPolicyOption() {
        var error: NSError?
        let maybeDb = LDBDatabase(path: path, error: &error,
            createIfMissing: true,
            bloomFilterBits: 10)
        if let error = error {
            XCTFail("Database.open failed with error: \(error)")
            return
        }
        let db = maybeDb!
        
        db["foo".UTF8] = "bar".UTF8
        
        XCTAssertEqual(db["foo".UTF8], "bar".UTF8)
    }
    
    func testCacheOption() {
        var error: NSError?
        let maybeDb = LDBDatabase(path: path, error: &error,
            createIfMissing: true,
            cacheCapacity: 2 << 20)
        if let error = error {
            XCTFail("Database.open failed with error: \(error)")
            return
        }
        let db = maybeDb!
        
        db["foo".UTF8] = "bar".UTF8
        
        XCTAssertEqual(db["foo".UTF8], "bar".UTF8)
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measureBlock() {
            // Put the code you want to measure the time of here.
        }
    }
    
}
